src/tools/fuzzing/heap-types.cpp

```diff
@@ -0,0 +1,499 @@
+/*
+ * Copyright 2021 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <variant>
+
+#include "tools/fuzzing/heap-types.h"
+#include "tools/fuzzing/parameters.h"
+
+namespace wasm {
+
+namespace {
+
+struct HeapTypeGenerator {
+  Random& rand;
+  FeatureSet features;
+  TypeBuilder builder;
+
+  // Map the HeapTypes we are building to their indices in the builder.
+  std::unordered_map<HeapType, Index> typeIndices;
+
+  // For each type we will build, the indices of its subtypes we will build.
+  std::vector<std::vector<Index>> subtypeIndices;
+
+  // Abstract over all the types that may be assigned to a builder slot.
+  using Assignable =
+    std::variant<HeapType::BasicHeapType, Signature, Struct, Array>;
+
+  // Top-level kinds, chosen before the types are actually constructed. This
+  // allows us to choose HeapTypes that we know will be subtypes of data or func
+  // before we actually generate the types.
```

Comment:
Knowing the kinds beforehand lets us pick subtypes of `HeapType::func` and `HeapType::data` that we haven't initialized yet. Otherwise we wouldn't know which of the uninitialized types would end up being valid subtypes. We don't need to differentiate between arrays and structs at this stage because `data` covers both of them.